symengine/functions.cpp

```diff
@@ -3142,4 +3142,200 @@ RCP<const Basic> abs(const RCP<const Basic> &arg)
     return make_rcp<const Abs>(arg);
 }
 
+template<class T>
+void get_unnested_args(const vec_basic& args, vec_basic& unnested_args)
+{
+    for (const auto &p: args) {
+        if (is_a<T>(*p)) {
+            get_unnested_args<T>(rcp_static_cast<const T>(p)->get_args(), unnested_args);
+        } else {
+            unnested_args.push_back(p);
+        }
+    }
+}
+
+Max::Max(const vec_basic&& arg)
+    :arg_{std::move(arg)}
+{
+    SYMENGINE_ASSERT(is_canonical(arg_))
+}
+
+bool Max::is_canonical(const vec_basic &arg) const
+{
+    if (arg.size() < 2)
+        return false;
+
+    bool non_number_exists = false;
+
+    for (const auto &p: arg) {
+        if (is_a<Complex>(*p) or is_a<Max>(*p))
+            return false;
+        if (not is_a_Number(*p))
+            non_number_exists = true;
+    }
+    return non_number_exists;   // all arguments cant be numbers
+}
+
+bool Max::__eq__(const Basic &o) const
+{
+    if (is_a<Max>(o) and
+        vec_basic_eq_perm(arg_, static_cast<const Max &>(o).arg_))
+        return true;
+    else
+        return false;
+}
+
+int Max::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<Max>(o))
+    const Max &s = static_cast<const Max &>(o);
+    if (arg_.size() != s.arg_.size())
+        return (arg_.size() < s.arg_.size()) ? -1 : 1;
+    return vec_basic_compare(arg_, s.arg_);
+}
+
+std::size_t Max::__hash__() const
+{
+    std::size_t seed = MAX;
+    for (const auto &p: arg_) {
+        hash_combine<Basic>(seed, *p);
+    }
+    return seed;
+}
+
+RCP<const Basic> max(const vec_basic &arg)
+{
+    bool number_set = false;
+    RCP<const Number> max_number, difference;
+    vec_basic unnested_args;
+    vec_basic non_number_args;
+
+    // Max inside Max should never occur! We unnest them recursively
+    get_unnested_args<Max>(arg, unnested_args);
```

Comment:
You don't need to unnest them recursively because you cannot have a `Max` inside a `Max`. One level of unnesting is enough